backend_kvm.go

```diff
@@ -0,0 +1,136 @@
+// +build linux
+// +build amd64
+
+package fakemachine
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+type KvmBackend struct {}
+
+func init() {
+	RegisterBackend(KvmBackend{})
+}
+
+func (b KvmBackend) Name() string {
+	return "kvm"
+}
+
+func (b KvmBackend) Supported() (bool, error) {
+	if _, err := os.Stat("/dev/kvm"); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (b KvmBackend) InitScriptModules() []string {
+	return []string{"virtio_pci", "virtio_console", "9pnet_virtio", "9p"}
+}
+
+func (b KvmBackend) InitScriptExtraCommands() []string {
+	return nil
+}
+
+func (b KvmBackend) NetworkdMatch() string {
+	return "e*"
+}
+
+func (b KvmBackend) CmdMountVolume(mount mountPoint) string {
+	return fmt.Sprintf("busybox mount -v -t 9p -o trans=virtio,version=9p2000.L,cache=loose,msize=262144 %s %s", mount.label, mount.machineDirectory)
+}
+
+func (b KvmBackend) GetFstabEntry(mount mountPoint) string {
+	return fmt.Sprintf("%s %s 9p trans=virtio,version=9p2000.L,cache=loose,msize=262144 0 0", mount.label, mount.machineDirectory)
+}
+
+func (b KvmBackend) GetImagePath(hostPath string, label string, index int) string {
+	return fmt.Sprintf("/dev/disk/by-id/virtio-%s", label)
+}
+
+func (b KvmBackend) GetJobOutputTTY(m *Machine) string {
+	// By default we send job output to the second virtio console,
+	// reserving /dev/ttyS0 for boot messages (which we ignore)
+	// and /dev/hvc0 for possible use by systemd as a getty
+	// (which we also ignore).
+	// If we are debugging, mix job output into the normal
+	// console messages instead, so we can see both.
+	if m.showBoot {
+		return "/dev/console"
+	}
+	return "/dev/hvc0"
+}
+
+func (b KvmBackend) Start(m *Machine) (int, error) {
```

Comment:
Similar to the above.. Makes me wonder whether it would make more sense to have a builder type pattern. Where the user to the backend creates one, configures it (set show boot, images and mounts points) and then queries its properties. Rather then just passing the whole machine and saying well good luck ;) (while show boot could in principle be changed behind its back between doing the GetJobOutputTTY and Start

Reply:
hmm i did think about start returning an array of the commands to run... but in some cases thats not enough, i.e libslirp helper for uml needs a sockerpair creating

Reply:
I think you misunderstood what i meant..

GetJobOutputTTY is basically; Please tell me what would the the tty *if* this particular option is set. With the assumption that that's the only determining factor. And ofcourse with the hidden contract that the showBoot option on the machine isn't changed.

What i'm wondering is if it wouldn't be nicer to build up a statefull backend object on which you can then query and just call start on. Which would also decouple the machine struct from the backends.. However a minimal variant could be something that results inthe follow pseudocode for the user:
```
instance := b.new_backend(machine);
tty := iinstance.GetJobOutputTTY();
imagepath := instance.GetImagePath(<image>);
instance.Start();
```
So again you'd flip things around; So where currently you ask  "imagine we'd start with <some useful options>" what would be your answer in that case and then only at Start() giving the full setup; With this change you'd start with give me an backend setup/instance for this configuration and then query things against that immutable instance.